File: ios/Packages/GemstoneServices/Sources/Device/DevicePlatform.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import enum Gemstone.Currency
import struct Gemstone.GemDeviceInfo
import protocol Gemstone.GemDeviceKeyServiceProtocol
import protocol Gemstone.GemDevicePlatform
import protocol Gemstone.GemPreferencesServiceProtocol
import GemstonePrimitives
import Keychain
import Primitives
import UIKit
import UserNotifications

public final class GemstoneDevicePlatform: GemDevicePlatform, @unchecked Sendable {
    private enum Keys {
        static let deviceId = "deviceId"
        static let deviceToken = "deviceToken"
        static let devicePrivateKey = "devicePrivateKey"
        static let devicePublicKey = "devicePublicKey"
    }

    private let preferencesService: any GemPreferencesServiceProtocol
    private let deviceKeyService: any GemDeviceKeyServiceProtocol
    private let keychain: any Keychain
    private let os: String
    private let model: String

    @MainActor
    public init(
        preferencesService: any GemPreferencesServiceProtocol,
        deviceKeyService: any GemDeviceKeyServiceProtocol,
        keychain: any Keychain = KeychainDefault(),
    ) {
        self.preferencesService = preferencesService
        self.deviceKeyService = deviceKeyService
        self.keychain = keychain
        os = UIDevice.current.osName
        model = UIDevice.current.modelName
    }

    public func deviceId() async throws -> String {
        try deviceKeyService.deviceId()
    }

    public func deviceInfo() async throws -> GemDeviceInfo {
        GemDeviceInfo(
            platform: Platform.ios.toGem(),
            platformStore: PlatformStore.current.toGem(),
            os: os,
            model: model,
            version: Bundle.main.releaseVersionNumber,
            localeIdentifier: Locale.current.identifier(.bcp47),
        )
    }

    public func pushToken() async throws -> String {
        try keychain.get(Keys.deviceToken) ?? .empty
    }

    public func setPushToken(_ token: String) throws {
        try keychain.accessibility(.whenUnlockedThisDeviceOnly, authenticationPolicy: []).set(token, key: Keys.deviceToken)
    }

    public func clearDeviceEntries() throws {
        for key in [Keys.deviceId, Keys.deviceToken, Keys.devicePrivateKey, Keys.devicePublicKey] {
            try keychain.remove(key)
        }
    }

    public func isPushEnabled() async throws -> Bool {
        guard preferencesService.isPushNotificationsEnabled() else {
            return false
        }
        return await UNUserNotificationCenter.current().notificationSettings().authorizationStatus.isAuthorized
    }

    public func getCurrency() async throws -> Gemstone.Currency {
        preferencesService.getCurrency()
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/device/DevicePlatform.kt
package com.gemwallet.android.data.services.gemstone.device

import com.gemwallet.android.ext.toGem
import android.content.Context
import androidx.core.app.NotificationManagerCompat
import com.gemwallet.android.application.device.cases.GetPushToken
import com.gemwallet.android.application.device.cases.RequestPushToken
import com.gemwallet.android.application.device.cases.SetPushToken
import com.gemwallet.android.ext.model
import com.gemwallet.android.ext.os
import com.gemwallet.android.model.NotificationsAvailable
import com.wallet.core.primitives.Platform
import com.wallet.core.primitives.PlatformStore
import uniffi.gemstone.GemDeviceInfo
import uniffi.gemstone.GemDevicePlatform
import uniffi.gemstone.GemDeviceKeyService
import java.util.Locale
import uniffi.gemstone.GemPreferencesService

class GemstoneDevicePlatform(
    private val context: Context,
    private val getPushToken: GetPushToken,
    private val setPushToken: SetPushToken,
    private val requestPushToken: RequestPushToken,
    private val platformStore: PlatformStore,
    private val notificationsAvailable: NotificationsAvailable,
    private val versionName: String,
    private val deviceKeyService: GemDeviceKeyService,
    private val preferencesService: GemPreferencesService,
) : GemDevicePlatform {

    override suspend fun deviceId(): String = deviceKeyService.deviceId()

    override suspend fun deviceInfo(): GemDeviceInfo = GemDeviceInfo(
        platform = Platform.Android.toGem(),
        platformStore = platformStore.toGem(),
        os = Platform.os,
        model = Platform.model,
        version = versionName,
        localeIdentifier = Locale.getDefault().toLanguageTag(),
    )

    override suspend fun pushToken(): String {
        val token = getPushToken.getPushToken()
        if (token.isEmpty()) {
            requestPushToken.requestToken { requested ->
                if (requested.isNotEmpty()) {
                    setPushToken.setPushToken(requested)
                }
            }
        }
        return token
    }

    override suspend fun isPushEnabled(): Boolean =
        notificationsAvailable &&
            preferencesService.isPushNotificationsEnabled() &&
            NotificationManagerCompat.from(context).areNotificationsEnabled()

    override suspend fun getCurrency(): String = preferencesService.getCurrency()
}
